arch/arm/chunkset_neon.c

```diff
@@ -0,0 +1,132 @@
+    *chunk = vreinterpretq_u8_s16(vdupq_n_s16(*(int16_t *)from));
+}
+
+static inline void chunkmemset_4(uint8_t *from, chunk_t *chunk) {
+    *chunk = vreinterpretq_u8_s32(vdupq_n_s32(*(int32_t *)from));
+}
+
+static inline void chunkmemset_8(uint8_t *from, chunk_t *chunk) {
+    *chunk = vcombine_u8(vld1_u8(from), vld1_u8(from));
+}
+
+#define CHUNKSIZE        chunksize_neon
+#define CHUNKCOPY        chunkcopy_neon
+#define CHUNKCOPY_SAFE   chunkcopy_safe_neon
+#define CHUNKUNROLL      chunkunroll_neon
+#define CHUNKMEMSET      chunkmemset_neon
+#define CHUNKMEMSET_SAFE chunkmemset_safe_neon
+
+uint8_t* CHUNKCOPY(uint8_t *out, uint8_t const *from, unsigned len);
+uint8_t* CHUNKUNROLL(uint8_t *out, unsigned *dist, unsigned *len);
+
+static inline uint8_t *chunkmemset_3(uint8_t *out, uint8_t *from, unsigned dist, unsigned len) {
+    uint8x8x3_t chunks;
+    unsigned sz = sizeof(chunks);
+    if (len < sz) {
+        out = CHUNKUNROLL(out, &dist, &len);
+        return CHUNKCOPY(out, out - dist, len);
+    }
+
+    /* Load 3 bytes 'a,b,c' from FROM and duplicate across all lanes:
+       chunks[0] = {a,a,a,a,a,a,a,a}
+       chunks[1] = {b,b,b,b,b,b,b,b}
+       chunks[2] = {c,c,c,c,c,c,c,c}. */
+    chunks = vld3_dup_u8(from);
+
+    unsigned rem = len % sz;
+    len -= rem;
+    while (len) {
+        /* Store "a,b,c, ..., a,b,c". */
+        vst3_u8(out, chunks);
+        out += sz;
+        len -= sz;
+    }
+
+    if (!rem)
+        return out;
+
+    /* Last, deal with the case when LEN is not a multiple of SZ. */
+    out = CHUNKUNROLL(out, &dist, &rem);
+    return CHUNKCOPY(out, out - dist, rem);
+}
+
+#if defined(__aarch64__) || defined(_M_ARM64)
+
+#define HAVE_CHUNKMEMSET_6
+
+static inline uint8_t *chunkmemset_6(uint8_t *out, uint8_t *from, unsigned dist, unsigned len) {
+    uint16x8x3_t chunks;
+    unsigned sz = sizeof(chunks);
+    if (len < sz) {
+        out = CHUNKUNROLL(out, &dist, &len);
+        return CHUNKCOPY(out, out - dist, len);
+    }
+
+    /* Load 6 bytes 'ab,cd,ef' from FROM and duplicate across all lanes:
+       chunks[0] = {ab,ab,ab,ab,ab,ab,ab,ab}
+       chunks[1] = {cd,cd,cd,cd,cd,cd,cd,cd}
+       chunks[2] = {ef,ef,ef,ef,ef,ef,ef,ef}. */
+    chunks = vld3q_dup_u16((unsigned short *)from);
+
+    unsigned rem = len % sz;
+    len -= rem;
+    while (len) {
+        /* Store "ab,cd,ef, ..., ab,cd,ef". */
+        vst3q_u16((unsigned short *)out, chunks);
+        out += sz;
+        len -= sz;
+    }
+
+    if (!rem)
+        return out;
+
+    /* Last, deal with the case when LEN is not a multiple of SZ. */
+    out = CHUNKUNROLL(out, &dist, &rem);
+    return CHUNKCOPY(out, out - dist, rem);
+}
+
+#endif
+
+static inline void loadchunk(uint8_t const *s, chunk_t *chunk) {
+    *chunk = *(chunk_t *)s;
+}
+
+static inline void storechunk(uint8_t *out, chunk_t *chunk) {
+#ifdef _MSC_VER
+    /* Cast to chunk_t pointer to avoid compiler error on MSVC ARM */
+    chunk_t *target = (chunk_t *)chunk;
```

Comment:
Sorry if you already tried this, but couldn't you unconditionally do memcpy((chunk_t *)out, chunk, sizeof(chunk_t))?

Also, shouldn't that be target = (chunk_t *)out ?  If that is indeed a thinko, that's an argument for just doing the unconditional cast...

Reply:
You are correct that it should be ```out``` instead of ```chunk```.

Reply:
next question: why didn't the test suite catch this?

Reply:
QEMU only works for Linux binaries.

Reply:
QEMU only works for Linux binaries.

Reply:
I was hoping there was a CI provider somewhere that made win10 arm support easy, but no.
(Mozilla does it, but I guess they had to roll their own a bit: https://wiki.mozilla.org/CI_Automation/windows10_aarch64 )

Reply:
It will only be a matter of time before GitHub supports ARM especially with the recent macOS announcements.

Reply:
Fixed.